CFC_DataCollector/modeinfer/pipeline.py

```diff
@@ -163,51 +202,69 @@ def getGroundTruthMode(self, section):
+#TODO: pass appropriate vector (featureMatrix[i]) rather than Matrix and index..
   def updateFeatureMatrixRowWithSection(self, featureMatrix, i, section):
-    featureMatrix[i, 0] = section['distance']
-    featureMatrix[i, 1] = (section['section_end_datetime'] - section['section_start_datetime']).total_seconds()
-
-    # Deal with unknown modes like "airplane"
-    try:
-      featureMatrix[i, 2] = section['mode']
-    except ValueError:
-      featureMatrix[i, 2] = 0
-
-    featureMatrix[i, 3] = section['section_id']
-    featureMatrix[i, 4] = calAvgSpeed(section)
-    speeds = calSpeeds(section)
-    if speeds != None:
-        featureMatrix[i, 5] = np.mean(speeds)
-        featureMatrix[i, 6] = np.std(speeds)
-        featureMatrix[i, 7] = np.max(speeds)
-    else:
-        # They will remain zero
-        pass
-    accels = calAccels(section)
-    if accels != None and len(accels) > 0:
-        featureMatrix[i, 8] = np.max(accels)
-    else:
-        # They will remain zero
-        pass
-    featureMatrix[i, 9] = ('commute' in section) and (section['commute'] == 'to' or section['commute'] == 'from')
-    featureMatrix[i, 10] = calHCR(section)
-    featureMatrix[i, 11] = calSR(section)
-    featureMatrix[i, 12] = calVCR(section)
-    if 'section_start_point' in section and section['section_start_point'] != None:
-        startCoords = section['section_start_point']['coordinates']
-        featureMatrix[i, 13] = startCoords[0]
-        featureMatrix[i, 14] = startCoords[1]
-    
-    if 'section_end_point' in section and section['section_end_point'] != None:
-        endCoords = section['section_end_point']['coordinates']
-        featureMatrix[i, 15] = endCoords[0]
-        featureMatrix[i, 16] = endCoords[1]
-    
-    featureMatrix[i, 17] = section['section_start_datetime'].time().hour
-    featureMatrix[i, 18] = section['section_end_datetime'].time().hour
-    
-    featureMatrix[i, 19] = mode_start_end_coverage(section, self.bus_cluster,105)
-    featureMatrix[i, 20] = mode_start_end_coverage(section, self.train_cluster,600)
+    points = section["track_points"]
+    if len(points) > 4:
+        points = np.array([[(datetime.datetime.strptime(point['time'].split('-')[0], "%Y%m%dT%H%M%S") - datetime.datetime(1970,1,1)).total_seconds(), 
+            point["track_location"]["coordinates"][0], point["track_location"]["coordinates"][1]] for point in points])
+        time_stamp = points[:,0].reshape(len(points[:,0]), 1)
+        lon_lat = points[:,1:]
+        model_ransac = linear_model.RANSACRegressor(linear_model.LinearRegression())
+        model_ransac.fit(lon_lat, time_stamp)
+        inlier_mask = model_ransac.inlier_mask_
+        outlier_mask = np.logical_not(inlier_mask)
+        print "total size: " + str(len(outlier_mask))
+        remove = [index for index,v in enumerate(outlier_mask) if v]
+        section["track_points"] = [v for j,v in enumerate(section["track_points"]) if j not in frozenset(remove)]
```

Comment:
One more comment: We should retain the initial (unsmoothed) data points separately to make it easier to explore alternate approaches. Two potential solutions:
- stick them into a separate "unsmoothed" collection, indexed by a shared trip + section ID
- stick them into a separate field of the same collection

           section['unsmoothed_track_points'] = section['track_points']
           section['track_points'] = [v for j,v in enumerate(section["track_points"]) if j not in frozenset(remove)]